storage/filesystem/dotgit/dotgit.go

```diff
@@ -1105,10 +1107,24 @@ func (d *DotGit) Module(name string) (billy.Filesystem, error) {
 	return d.fs.Chroot(d.fs.Join(modulePath, name))
 }
 
+func (d *DotGit) AddAlternate(remote string) error {
+	altpath := d.fs.Join(objectsPath, infoPath, alternatesPath)
+	f, err := d.fs.OpenFile(altpath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0640)
+	if err != nil {
+		return err
+	}
+	line := path.Join(remote, objectsPath) + "\n"
+	_, err = io.WriteString(f, line)
```

Comment:
We should align with [upstream](https://github.com/git/git/blob/3a06386e314565108ad56a9bdb8f7b80ac52fb69/object-file.c#L651) and try to acquire a lock before making changes to the file. The go-billy filesystem abstraction already supports it, so please add a call to `f.Lock()` and defer the `f.Unlock()`.